File: backend/model/product.go
package model

import "time"

// Product 商品模型
type Product struct {
	ID           int64     `json:"id" gorm:"primaryKey"`
	Title        string    `json:"title"`
	Description  string    `json:"description"`
	Price        float64   `json:"price"`
	CategoryID   int64     `json:"categoryId"`
	ConditionID  int64     `json:"conditionId"`
	SellerID     int64     `json:"sellerId"`
	Status       string    `json:"status"`
	MainImageURL string    `json:"mainImageUrl" gorm:"column:main_image_url"`
	CreatedAt    time.Time `json:"createdAt"`
	UpdatedAt    time.Time `json:"updatedAt"`
}

// ProductImage 商品图片模型
type ProductImage struct {
	ID        int64  `json:"id" gorm:"primaryKey"`
	ProductID int64  `json:"productId"`
	URL       string `json:"url"`
	IsPrimary bool   `json:"isPrimary"`
	SortOrder int    `json:"sortOrder"`
}

// ProductDetailDTO 商品详情DTO
type ProductDetailDTO struct {
	ID             int64          `json:"id"`
	Title          string         `json:"title"`
	Description    string         `json:"description"`
	Price          float64        `json:"price"`
	CategoryID     int64          `json:"categoryId"`
	ConditionID    int64          `json:"conditionId"`
	ConditionName  string         `json:"conditionName"`
	MainImageURL   string         `json:"mainImageUrl"`
	Images         []ProductImage `json:"images"`
	TagIDs         []int64        `json:"tagIds"`
	Seller         SellerInfo     `json:"seller"`
	ViewerIsSeller bool           `json:"viewerIsSeller"`
	Status         string         `json:"status"`
	CreatedAt      time.Time      `json:"createdAt"`
	UpdatedAt      time.Time      `json:"updatedAt"`
	SellerWechat   *string        `json:"sellerWechat,omitempty"`
}

// ProductCardDTO 商品卡片DTO
type ProductCardDTO struct {
	ID          int64     `json:"id"`
	Title       string    `json:"title"`
	Price       float64   `json:"price"`
	MainImage   string    `json:"mainImageUrl"`
	Status      string    `json:"status"`
	SellerID    int64     `json:"sellerId"`
	CategoryID  int64     `json:"categoryId"`
	ConditionID int64     `json:"conditionId"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
}

// SellerInfo 卖家简要信息
type SellerInfo struct {
	ID        int64  `json:"id"`
	Nickname  string `json:"nickname"`
	AvatarUrl string `json:"avatarUrl"`
}

File: frontend/src/api/product.ts
import request from '@/utils/request'
import type { ApiResponse, PageResult } from '@common/types/api'
import type { Product } from '@common/types/product'
import type { ProductCondition } from '@common/types/product_condition'

export interface ProductImage {
  id: number
  url: string
  sortOrder: number
  isPrimary: boolean
}

export interface ProductDetail extends Product {
  conditionName: string
  images: ProductImage[]
  tagIds: number[]
  seller: {
    id: number
    nickname: string
    avatarUrl: string
  }
  viewerIsSeller: boolean
  sellerWechat: string | null
}

// 发布商品参数 (FormData)
// title, description, price, categoryId, tagIds, conditionId, images
// 这里只定义接口，实际调用时传 FormData

// 编辑商品参数
export interface UpdateProductParams {
  title?: string
  description?: string
  price?: number
  categoryId?: number
  tagIds?: number[] // 前端使用数组，在请求前转换为逗号分隔字符串
  conditionId?: number
  imageUrls?: string[]
}

// 搜索参数
export interface ProductSearchParams {
  q?: string
  categoryId?: number
  tagIds?: string
  conditionIds?: string
  minPrice?: number
  maxPrice?: number
  publishedTimeRange?: string
  sort?: string
  page?: number
  pageSize?: number
}

// 状态变更参数
export interface ProductStatusParams {
  action: 'delist' | 'relist' | 'sold'
}

// 联系卖家响应
export interface ContactSellerResponse {
  canContact: boolean
  sellerWechat?: string
  tips?: string
}

export function createProduct(data: FormData) {
  return request.post<ApiResponse<Product>>('/products', data, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })
}

export function updateProduct(id: number, data: UpdateProductParams) {
  // 转换 tagIds 数组为逗号分隔字符串
  const requestData = {
    ...data,
    tagIds:
      data.tagIds && Array.isArray(data.tagIds) ? (data.tagIds.join(',') as any) : data.tagIds,
  }
  return request.put<ApiResponse<Product>>(`/products/${id}`, requestData)
}

export function changeProductStatus(id: number, data: ProductStatusParams) {
  return request.post<ApiResponse<Product>>(`/products/${id}/status`, data)
}

export function undoProductStatusChange(id: number) {
  return request.post<ApiResponse<Product>>(`/products/${id}/status/undo`)
}

export function getProductDetail(id: number) {
  return request.get<ApiResponse<ProductDetail>>(`/products/${id}`)
}

export function recordProductView(id: number) {
  return request.post<ApiResponse<{ recorded: boolean }>>(`/products/${id}/view`)
}

export function getMyProducts(params: { keyword?: string; page?: number; pageSize?: number }) {
  return request.get<ApiResponse<PageResult<Product>>>('/products/my', { params })
}

export function searchProducts(params: ProductSearchParams) {
  return request.get<ApiResponse<PageResult<Product>>>('/products/search', { params })
}

export function getProductsByCategory(categoryId: number, params: ProductSearchParams) {
  return request.get<ApiResponse<PageResult<Product>>>(`/products/category/${categoryId}`, {
    params,
  })
}

export function getProductContact(id: number) {
  return request.get<ApiResponse<ContactSellerResponse>>(`/products/${id}/contact`)
}

export function getProductConditions() {
  return request.get<ApiResponse<ProductCondition[]>>('/product-conditions')
}
